modules/build/src/main/scala/scala/build/bsp/BspServerProxy.scala

```diff
@@ -0,0 +1,126 @@
+          else CompletableFuture.completedFuture(res)) match {
+          case Left(errorMessage) =>
+            CompletableFuture.completedFuture(
+              responseError(s"Workspace reload failed, couldn't load sources: $errorMessage")
+            )
+          case Right(r) => r
+        }
+      else CompletableFuture.completedFuture(
+        responseError(
+          s"Workspace reload failed, inputs file missing from workspace directory: ${ideInputsJsonPath.toString()}"
+        )
+      )
+    }
+
+  private def reloadBsp(
+    previousInputs: Inputs,
+    newInputs: Inputs
+  ): CompletableFuture[AnyRef] = {
+    val previousTargetIds = currentBspServer.targetIds
+    currentBloopCompiler = createBloopCompiler(newInputs)
+    currentBspServer = createBspServer(currentBloopCompiler, newInputs)
+    val newTargetIds = currentBspServer.targetIds
+    prepareBuild() match {
+      case Left((buildException, scope)) =>
+        CompletableFuture.completedFuture(
+          new ResponseError(
+            JsonRpcErrorCodes.InternalError,
+            s"Can't reload workspace, build failed for scope: ${scope.name}: ${buildException.message}",
+            new Object()
+          )
+        )
+      case Right(preBuildProject) =>
+        if (previousInputs.projectName != preBuildProject.mainScope.project.projectName) {
+          val events = newTargetIds.map(buildTargetIdToEvent(_, b.BuildTargetEventKind.CREATED)) ++
+            previousTargetIds.map(buildTargetIdToEvent(_, b.BuildTargetEventKind.DELETED))
+          val didChangeBuildTargetParams = new b.DidChangeBuildTarget(events.asJava)
+          currentBspServer.client.foreach(_.onBuildTargetDidChange(didChangeBuildTargetParams))
+        }
+        CompletableFuture.completedFuture(new Object())
+    }
+  }
+
+  private def createBloopCompiler(inputs: Inputs): BloopCompiler = {
+    val bloopServer = BloopServer.buildServer(
+      bloopRifleConfig,
+      "scala-cli",
+      Constants.version,
+      (inputs.workspace / Constants.workspaceDirName).toNIO,
+      Build.classesRootDir(inputs.workspace, inputs.projectName).toNIO,
+      localClient,
+      threads.buildThreads.bloop,
+      logger.bloopRifleLogger
+    )
+    new BloopCompiler(
+      bloopServer,
+      20.seconds,
+      strictBloopJsonCheck = buildOptions.internal.strictBloopJsonCheckOrDefault
+    )
+  }
+
+  private def createBspServer(bloopCompiler: BloopCompiler, inputs: Inputs) =
+    new BspServer(bloopCompiler.bloopServer.server, compile, logger, inputs)
+
+  private def buildTargetIdToEvent(
+    targetId: b.BuildTargetIdentifier,
+    eventKind: b.BuildTargetEventKind
+  ): b.BuildTargetEvent = {
+    val event = new b.BuildTargetEvent(targetId)
+    event.setKind(eventKind)
+    event
+  }
+
+  private def responseError(
+    message: String,
+    errorCode: Int = JsonRpcErrorCodes.InternalError
+  ): ResponseError =
+    new ResponseError(errorCode, message, new Object())
```

Comment:
even though we return a valid error response to the `workspace/reload` request, it doesn't seem like `IntelliJ` is handling it properly, as no error message is shown in the UI. We may consider sending a BSP `logMessage` request, so the `IntelliJ` user has at least a slight chance of noticing that something went wrong (it'd show in the BSP build log, not just the BSP logs in the IntelliJ directory). `showMessage` isn't an option either, as `IntelliJ` ignores that as well. TBD, I'm open to suggestions (although I'd rather do it in a separate PR as well)